docs/design/graceful-domain-failover/3051-graceful-domain-failover.md

```diff
@@ -0,0 +1,230 @@
+3. The domain state has to be stable prior to graceful failover.
+
+## Limitation
+No concurrent graceful domain failover will be supported.
+
+## Proposal
+
+Due to the complexity of the protocol, it will go through the architecture from cross-cluster level, single cluster level to host level.
+
+The basic protocol is to insert markers in the active-to-passive cluster to indicate the boundary when the domain switches to passive. On the other side, the passive-to-active cluster listens to those markers and switches domain to active after receiving all the markers.
+
+### Cross cluster level
+
+![cross clusters sequence diagram](3051-cross-clusters.png)
+1, The operator issues a graceful failover to the passive-to-active cluster.
+
+2 - 3, The passive-to-active cluster gets domain data from all other clusters for two purposes: 1. Make sure the network and clusters are available prior to start the graceful failover. 2. Make sure there is no ongoing failover.
+
+4 - 5, If the check fails, return an error to the operator indicating the graceful failover abort.
+
+6, After the graceful failover is initiated, cluster Y updates the domain to pending_active with a higher failover version to database.
+
+7, Respond the operator indicating the graceful failover initiated.
+
+8, The domain update event in step 6 replicates to cluster X.
+
+9, Cluster X updates the domain with the higher failover version and sets the domain to passive.
+
+10, Each shard receives a domain failover notification. The shard persists the pending failover marker and inserts the marker to the replication queue.
+
+11, The inserted failover marker replicates to cluster Y.
+
+12, Each shard in cluster Y listens to the failover marker and reports the ‘ready’ state to the failover coordinator after it receives the failover marker.
+
+13, The failover coordinator updates domain from pending_active to active when received ‘ready’ signal from all shards.
+
+14, The failover coordinator updates domain from pending_active to active when the timeout hits and regardless how many ‘ready’ signals it received.
+
+From the high level sequence diagram, it explains how the protocol works within multi-clusters. There is detail at cluster level.
+
+### Cluster X
+![cross cluster X sequence diagram](3051-clusterX.png)
+
+1. Frontend receives a domain replication message.
+2. Frontend updates the domain data in Database with activeCluster set to Cluster Y and a higher failover version.
+3. Domain cache fetch domain updates in a refresh loop.
+4. Database returns the domain data.
+5. After the domain updates, domain cache sends a domain failover notification to each shard.
+6. After the domain updates, domain cache sends a domain failover notification to each shard.
+7. Shard 1 updates the shard info with a pending failover marker to insert.
+8. Shard 1 try to insert the failover marker and remove the pending failover marker from shard info after successful insertion.
```

Comment:
Yes. Added a statement to explicitly mentioned in the task processor section that it pauses on processing tasks during the Pending_Active. This pause function relies on the multi-tenancy design so that it won't block other domains.

Reply:
It is not just task processor. All start/signal, etc requests should be rejected if ever reach this shard after the failover marker is pending.

Reply:
Yes. The requests will be rejected after the marker inserted.